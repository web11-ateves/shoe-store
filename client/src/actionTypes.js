const constants = {
  NEW_INVENTORY_ARRIVED: 'NEW_INVENTORY_ARRIVED',
  ADD_INVENTORY: 'ADD_INVENTORY',
  UPDATE_INVENTORY: 'UPDATE_INVENTORY',
  TOGGLE_SUGGESTION: 'TOGGLE_SUGGESTION',
  CHECK_TRANSFER: 'CHECK_TRANSFER',
  SUGGEST_TRANSFER: 'SUGGEST_TRANSFER',
  CONFIRM_INVENTORY_TRANSFER: 'CONFIRM_INVENTORY_TRANSFER',
  RESET_TRANSFER: 'RESET_TRANSFER',
  UPDATE_ALERT: 'UPDATE_ALERT',
  UPDATE_QUERY: 'UPDATE_QUERY',
  UPDATE_TIMESTAMP: 'UPDATE_TIMESTAMP',
};

export default constants;
