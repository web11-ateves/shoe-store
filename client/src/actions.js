import { v4 } from 'uuid';
import types from './actionTypes';

export const newInventoryArrived = (store, product, stock, transfer = false) => ({
  type: types.NEW_INVENTORY_ARRIVED,
  store,
  product,
  stock,
  transfer,
});

export const addInventory = (store, product, stock, timestamp) => ({
  type: types.ADD_INVENTORY,
  id: v4(),
  store,
  product,
  stock,
  timestamp,
});

export const updateInventory = (id, stock, timestamp) => ({
  type: types.UPDATE_INVENTORY,
  id,
  stock,
  timestamp,
});

export const toggleSuggestion = (status) => ({
  type: types.TOGGLE_SUGGESTION,
  status,
});

export const suggestTransfer = (fromId, fromStock, toId, toStock, message) => ({
  type: types.SUGGEST_TRANSFER,
  fromId,
  fromStock,
  toId,
  toStock,
  message,
});

export const confirmTransfer = (payload) => ({
  type: types.CONFIRM_INVENTORY_TRANSFER,
  fromId: payload.fromId,
  fromStock: payload.fromStock,
  toId: payload.toId,
  toStock: payload.toStock,
});

export const resetTransfer = () => ({
  type: types.RESET_TRANSFER,
  fromId: '',
  fromStock: '',
  toId: '',
  toStock: '',
  message: '',
});

export const updateAlert = (message, severity) => ({
  type: types.UPDATE_ALERT,
  message,
  severity,
});

export const updateQuery = (storeQuery, productQuery) => ({
  type: types.UPDATE_QUERY,
  storeQuery,
  productQuery,
});

export const updateTimestamp = (timestamp) => ({
  type: types.UPDATE_TIMESTAMP,
  timestamp,
});
