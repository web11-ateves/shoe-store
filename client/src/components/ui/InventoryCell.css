.inventory-enter {
    opacity: 0.1;
}
.inventory-enter-active {
    opacity: 1;
    background-color: #FFD499;
    transition: opacity .5s ease-in;
}