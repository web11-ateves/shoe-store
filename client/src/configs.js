export const warningLevel = 20;
export const criticalLevel = 5;
