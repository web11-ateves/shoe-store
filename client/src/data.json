{
	"stores": [
		"Centre Eaton",
		"Destiny USA Mall",
		"Pheasant Lane Mall",
		"Holyoke Mall", "Maine Mall",
		"Crossgates Mall",
		"Burlington Mall",
		"Solomon Pond Mall",
		"Auburn Mall",
		"Waterloo Premium Outlets"
	],
	"models": [
		"ADERI",
		"MIRIRA",
		"CAELAN",
		"BUTAUD",
		"SCHOOLER",
		"SODANO",
		"MCTYRE",
		"CADAUDIA",
		"RASIEN",
		"WUMA",
		"GRELIDIEN",
		"CADEVEN",
		"SEVIDE",
		"ELOILLAN",
		"BEODA",
		"VENDOGNUS",
		"ABOEN",
		"ALALIWEN",
		"GREG",
		"BOZZA"
	]
}