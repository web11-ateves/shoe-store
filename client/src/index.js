import React from 'react';
import { render } from 'react-dom';
import { Provider } from 'react-redux';
import { w3cwebsocket as W3CWebSocket } from 'websocket';
import store from './store';
import App from './components/App';
import { updateAlert, newInventoryArrived } from './actions';

window.React = React;
window.store = store;

const webSocketConnect = () => {
  let timerId;
  const client = new W3CWebSocket('ws://localhost:8080/');
  client.onopen = () => {
    console.log('WebSocket connected successfully');
    store.dispatch(updateAlert('WebSocket connected successfully', 'success'));
    clearInterval(timerId);
  };
  client.onmessage = (event) => {
    const data = JSON.parse(event.data);
    console.log(data);
    store.dispatch(newInventoryArrived(data.store, data.model, data.inventory));
  };
  client.onerror = (err) => {
    console.error('WebSocket encountered error: ', err.message, 'Closing socket');
    client.close();
  };
  client.onclose = () => {
    console.log('WebSocket closed. Reconnect will be attempted in 1 second.');
    store.dispatch(updateAlert('WebSocket closed. Reconnect will be attempted in 1 second.', 'error'));
    timerId = setTimeout(() => {
      webSocketConnect();
    }, 1000);
  };
  return client;
};

webSocketConnect();

render(
  <Provider store={store}>
    <App />
  </Provider>,
  document.getElementById('root'),
);
