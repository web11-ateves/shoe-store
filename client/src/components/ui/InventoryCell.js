/* eslint-disable no-nested-ternary */
import React from 'react';
import { CSSTransition } from 'react-transition-group';
import PropTypes from 'prop-types';
import './InventoryCell.css';
import StyledTableCell from '../hoc/StyledTableCell';
import { criticalLevel, warningLevel } from '../../configs';

const InventoryCell = ({ value, highlight }) => (
  <CSSTransition in={highlight} timeout={500} classNames="inventory">
    <StyledTableCell
      align="center"
      className={value <= criticalLevel
        ? 'critical-inventory'
        : value <= warningLevel
          ? 'warning-inventory'
          : 'high-inventory'}
    >
      {value || ''}

    </StyledTableCell>
  </CSSTransition>
);

InventoryCell.propTypes = {
  value: PropTypes.number,
  highlight: PropTypes.bool,
};

InventoryCell.defaultProps = {
  value: 0,
  highlight: false,
};

export default InventoryCell;
