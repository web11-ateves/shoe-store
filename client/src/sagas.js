import {
  put, takeEvery, all, select,
} from 'redux-saga/effects';
import types from './actionTypes';
import { criticalLevel, warningLevel } from './configs';
import {
  newInventoryArrived, updateInventory, suggestTransfer,
  updateAlert, updateTimestamp, resetTransfer,
} from './actions';
import {
  findByAttr, findByAttrs, getFirstArrayItem,
  getLastArrayItem, filterArrayByAttrs, sortBy,
} from './lib/helpers';

export const getInventories = (state) => state.inventories;
export const getTransfer = (state) => state.transfer;
export const getSuggestion = (state) => state.suggestion;

function* changeInventory(action) {
  const inventories = yield select(getInventories);
  const transfer = yield select(getTransfer);
  const suggestion = yield select(getSuggestion);

  if (action.stock <= warningLevel) {
    yield put(updateAlert(`Inventory is low for [${action.store}, ${action.product}]`, 'warning'));
  }

  const timestamp = Date.now();
  yield put(updateTimestamp(timestamp));

  let inventory;
  if (action.id) {
    inventory = findByAttr(inventories, 'id', action.id);
  } else {
    inventory = findByAttrs(inventories,
      { attr: 'store', op: 'like', value: [action.store] },
      { attr: 'product', op: 'like', value: [action.product] });
  }
  yield put(updateInventory(inventory.id, action.stock, timestamp));

  if ((transfer.fromId === inventory.id || transfer.toId === inventory.id)) {
    yield put(resetTransfer());
  }

  if (suggestion) {
    yield put({
      type: types.CHECK_TRANSFER,
      product: action.product,
    });
  }
}

function* checkInventoryTransfer(action) {
  const inventories = yield select(getInventories);
  const transfer = yield select(getTransfer);

  const sortedInventories = filterArrayByAttrs(inventories,
    { attr: 'product', op: 'like', value: action.product },
    { attr: 'timestamp', op: 'not', value: '' })
    .sort(sortBy('number', 'stock'));

  const fromInventory = getFirstArrayItem(sortedInventories);
  const toInventory = getLastArrayItem(sortedInventories);

  const stockToTransfer = warningLevel - toInventory.stock + 1;

  if (transfer.message === '' // disregards active transfers
    && !!toInventory.timestamp // disregards initial inventory
    && toInventory.stock <= criticalLevel
    && fromInventory.stock - stockToTransfer > warningLevel) {
    const transferMessage = `Would you like to transfer [${stockToTransfer}] units of '${action.product}' from '${fromInventory.store}' to '${toInventory.store}'?`;
    console.log(transferMessage);
    yield put(suggestTransfer(
      fromInventory.id,
      fromInventory.stock - stockToTransfer,
      toInventory.id,
      toInventory.stock + stockToTransfer,
      transferMessage,
    ));
  }
}

function* confirmTransfer(action) {
  const inventories = yield select(getInventories);

  const fromInventory = findByAttr(inventories, 'id', action.fromId);
  fromInventory.stock = action.fromStock;

  const toInventory = findByAttr(inventories, 'id', action.toId);
  toInventory.stock = action.toStock;

  yield put(newInventoryArrived(
    fromInventory.store,
    fromInventory.product,
    action.fromStock,
    true,
  ));
  yield put(newInventoryArrived(
    toInventory.store,
    toInventory.product,
    action.toStock,
    true,
  ));
  yield put(resetTransfer());
}

function* watchAll() {
  yield all([
    takeEvery(types.NEW_INVENTORY_ARRIVED, changeInventory),
    takeEvery(types.CHECK_TRANSFER, checkInventoryTransfer),
    takeEvery(types.CONFIRM_INVENTORY_TRANSFER, confirmTransfer),
  ]);
}

export default watchAll;
