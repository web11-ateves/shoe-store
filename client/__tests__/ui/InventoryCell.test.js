/* eslint-disable no-undef */
import React from 'react';
import toJSON from 'enzyme-to-json';
import { compose } from 'redux';
import InventoryCell from '../../src/components/ui/InventoryCell';
import { criticalLevel, warningLevel } from '../../src/configs';

const { shallow, render } = Enzyme;

describe('<InventoryCell /> Component', () => {
  const shallowExpect = compose(expect, toJSON, shallow);

  const criticalInventory = criticalLevel - 1;
  const warningInventory = warningLevel - 1;
  const highInventory = 90;

  it('Renders correctly for critical inventory', () => shallowExpect(
    <InventoryCell value={criticalInventory} highlight />,
  ).toMatchSnapshot());

  it('Renders correctly for warning inventory', () => shallowExpect(
    <InventoryCell value={warningInventory} />,
  ).toMatchSnapshot());

  it('Renders correctly for high inventory', () => shallowExpect(
    <InventoryCell value={highInventory} />,
  ).toMatchSnapshot());

  it('Renders correct class for critical inventory', () => expect(
    render(<InventoryCell value={criticalInventory} highlight />)
      .hasClass('critical-inventory'),
  ).toBe(true));

  it('Renders correct class for low inventory', () => expect(
    render(<InventoryCell value={warningInventory} />)
      .hasClass('warning-inventory'),
  ).toBe(true));

  it('Renders correct class for high inventory', () => expect(
    render(<InventoryCell value={highInventory} />)
      .hasClass('high-inventory'),
  ).toBe(true));
});
