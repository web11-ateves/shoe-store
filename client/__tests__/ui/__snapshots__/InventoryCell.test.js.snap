// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<InventoryCell /> Component Renders correctly for critical inventory 1`] = `
<CSSTransition
  classNames="inventory"
  in={true}
  timeout={500}
>
  <WithStyles(WithStyles(ForwardRef(TableCell)))
    align="center"
    className="critical-inventory"
  >
    4
  </WithStyles(WithStyles(ForwardRef(TableCell)))>
</CSSTransition>
`;

exports[`<InventoryCell /> Component Renders correctly for high inventory 1`] = `
<CSSTransition
  classNames="inventory"
  in={false}
  timeout={500}
>
  <WithStyles(WithStyles(ForwardRef(TableCell)))
    align="center"
    className="high-inventory"
  >
    90
  </WithStyles(WithStyles(ForwardRef(TableCell)))>
</CSSTransition>
`;

exports[`<InventoryCell /> Component Renders correctly for warning inventory 1`] = `
<CSSTransition
  classNames="inventory"
  in={false}
  timeout={500}
>
  <WithStyles(WithStyles(ForwardRef(TableCell)))
    align="center"
    className="warning-inventory"
  >
    19
  </WithStyles(WithStyles(ForwardRef(TableCell)))>
</CSSTransition>
`;
